### ISSUE TITLE
Rails 7.1 - can't define enums for attributes which don't exist in table

### ISSUE THREAD

**Issue by ejskipper** [rails/rails#49717]
Not sure if this is an intended change or not. Pre 7.1, I could define an attr on a model using `attribute` and define enum values for it, then use enum helper methods. I'm using this heavily in my current project for i18ns etc, so defining the enum values as an array and manually validating won't cover all my use cases

Since 7.1, this isn't possible.

### Steps to reproduce

```ruby
# frozen_string_literal: true

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails", branch: "main"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end
end

class Post < ActiveRecord::Base
  attribute :topic
  enum topic: %i[sports politics entertainment]
end

class BugTest < Minitest::Test
  def test_association_stuff
    post = Post.create!
    post.assign_attributes(topic: :sports)

    assert_equal true, post.sports?
  end
end# frozen_string_literal: true


```

### Expected behavior
Post should instantiate correctly and can use enum helper methods

### Actual behavior
Error thrown

`RuntimeError: Unknown enum attribute 'topic' for Post`

### System configuration
**Rails version**: 7.1.1

**Ruby version**: ruby 3.1.2p20

**Comment by zzak** [comment]
Thanks for your report!

I've bisected this down to 6c5fab0668c1872fe827507f45ef400a20e8c646 which was introduced in #45734.

It looks like intended behavior, but it's unclear to me why exactly this was changed. /cc @fatkodima

**Comment by fatkodima** [comment]
Yes, that was intentional. See related issue https://github.com/rails/rails/issues/45668 for the discussion of the problem.

cc @byroot

**Comment by ejskipper** [comment]
Thanks for the response. I'm still a bit unclear on the rationale behind this change; looking at the reported behaviour in #45668 , it seems to me that we have the same behaviour with other types, not just enums, so not sure where the difference is. 
Eg if the column is a simple integer and we define `attribute :wrong_attributee` on Post model, we still will not throw on `new_post = Post.create!(wrong_attributee:  2) # column is not in the schema cache, so why no error?` and the Post will not be created with wrong_attributee = 2, just like in the code reported re: enum. (Happy to post a full test example if useful)

Either way though, I guess it boils down to the statement 

> While it is possible to have attributes without the columns (defined with the [attribute](https://api.rubyonrails.org/classes/ActiveRecord/Attributes/ClassMethods.html#method-i-attribute) method), the enum (which is also defined by attribute) should always have a corresponding column.

Perhaps we just need some documentation/explanation of this?

**Comment by jkwuc89** [comment]
I also have a project that is being significantly impacted by this. Is there a "simple/straightforward" way to fix the offending enum definitions in the models? Currently, this is preventing me from upgrading to Rails 7.1.

**Comment by byroot** [comment]
> Is there a "simple/straightforward" way to fix the offending enum definitions in the models?

It's not quite clear to me what you are asking. Invalid enums raise an error on boot, you should be able to catch them by running your test suite.

Or are you asking for something else?

**Comment by byroot** [comment]
I don't follow. You say:

> we define enum's whose values are used to populate a `character varying` column

And then:

> we can no longer define enum's inside models without a corresponding database table column.

So first you say the enum is backed by a column, then you say it's not. It's really confusing.

**Comment by jkwuc89** [comment]
My apologies...it looks like it may be a simple case of `order_by_types` not matching the `order_by_type` column name. I am going to check my `enum`'s and correct the name mismatches to see if that fixes things.

**Comment by jkwuc89** [comment]
Changing the `enum` name to match the corresponding column name is fixing things for me. Again...my apologies for the confusion/noise in this thread.

**Comment by ejskipper** [comment]
I think as jkwuc says, that wasn't actually related to my original question.

It seems like the my issue (enums for attributes which are not backed by a column) is now explicitly unsupported, and the way forwards is to rewrite the enums into arrays and explicitly validate inclusion in those values, and manually feed in the enum options and i18ns to the inputs on forms, etc. If anyone has elegant ways of doing this then would be interested to hear!

I guess the documentation for enum does say pretty clearly that it should be backed by a DB column, so hopefully that's sufficient for others going forwards to know that this use case isn't supported.
> Declare an enum attribute where the values map to integers in the database, but can be queried by name

**Comment by jonathanhefner** [comment]
Eventually I would like to port `ActiveRecord::Enum` to Active Model, so I think it would be nice to support non-column-backed attributes.  Otherwise, there would be a disparity between `enum` in an Active Record model and `enum` in an Active Model model.

Instead of checking for a database column, we could check that the attribute has a declared type.  In other words, it must either be defined in the schema or be declared with an explicit type via `ActiveRecord::Base.attribute`.

I've submitted #49769 to accomplish this.

**Comment by p8** [comment]
@jonathanhefner I was working on a PR to port ActiveRecord::Enum to ActiveModel some time ago.
I'll see if I can get that ready.

**Comment by ejskipper** [comment]
Looks like a great solution, thanks for the PR @jonathanhefner and for the continued support @byroot 🙏

**Comment by jasonl** [comment]
We have also encountered this issue, but in a slightly different way. We have a model where we added an enum column to it in a migration created long after the original table. This now creates the problem where we cannot run our migrations for an empty database, as the model defines the enum, but the column backing it doesn't exist in the DB, and won't until the migration is run which creates it.

We've hacked around it by defining an empty AR model for the table in the migrations which break, but this seems inelegant to say the least.

**Comment by skipkayhil** [comment]
> We have also encountered this issue, but in a slightly different way. We have a model where we added an enum column to it in a migration created long after the original table. This now creates the problem where we cannot run our migrations for an empty database, as the model defines the enum, but the column backing it doesn't exist in the DB, and won't until the migration is run which creates it.
> 
> We've hacked around it by defining an empty AR model for the table in the migrations which break, but this seems inelegant to say the least.

Can you elaborate on why the model is being loaded during migrations? I'm not sure if that's a common case

**Comment by deepakmahakale** [comment]
I am facing this same issue when the model is used in some migration before the enum is introduced for example

```ruby
# app/models/user.rb
class User < ApplicationEnhancedVersionedRecord
  enum status: { draft: 'draft', published: 'published' }, _default: :draft
end

# 20220101000000_add_posts
class AddPosts < ActiveRecord::Migration[7.0]
  def change
    create_table :posts, id: :uuid do |t|
      t.string   :title, null: false
    end
  end
end

# 20220102000000_some_data_migration
class AddStatusToPosts < ActiveRecord::Migration[7.0]
  def change
    Post.find_each {...}
  end
end

# 20220103000000_add_status_to_posts
class AddStatusToPosts < ActiveRecord::Migration[7.0]
  def change
    add_column :posts, :status, :string, default: "draft"
  end
end
```

Now I get 

```ruby
StandardError: An error has occurred, this and all later migrations canceled: (StandardError)

Unknown enum attribute 'status' for Post
```

**Comment by FedorenkoCodes** [comment]
I have issue similar to the one @jasonl has. There is an 3 years old migration in my app, which adds a column to User database table, and it populates it in the loop for each user:
```
def up
    add_column :users, :new_filled_column, :string

    User.find_each do |user|
      user.update(new_filled_column: User.default_field_value)
    end
  end
```
    
A year later another migration adds new column for storing enum value
```
add_column :users, :another_enum_column, :integer
```
and corresponding code to set its values
```
enum another_enum_column: { value_one: 1, value_two: 2 }, _prefix: true
```

So now when I reset my schema, I get an error: "`Unknown enum attribute 'another_enum_column' for User`", because column `another_enum_column` doesn't yet exist when older migration calls User model.

Is it possible you might have some kind of solution to resolve this error? I really don't like the idea to edit old migrations.

**Comment by byroot** [comment]
> @byroot Is it possible you might have some kind of solution to resolve this error? I really don't like the idea to edit old migrations.

You probably won't like my answer but:

  - Accessing models in migration is an anti-pattern precisely because it introduce this kind of issues.
  - There is absolutely no reason to run 3 years old migrations. Use `db:schema:load` to directly start from a recent state.

Old migrations are meant to be deleted.

**Comment by jfloff** [comment]
I am getting this error while including an enum through a concern.

In my concern I have: `enum id_verification_status: { ... }`

**Comment by jonathanhefner** [comment]
@jfloff Can you open a new issue with a minimal reproduction script based on https://github.com/rails/rails/blob/main/guides/bug_report_templates/active_record_gem.rb?

**Comment by arcreative** [comment]
@byroot Not sure I'm a fan of the absolutism here. 3 year old migrations could be run for any number of reasons, and the fact that we're having this conversation is proof of that. As part of a major overhaul, I also would like to be able to run migrations from scratch and ensure they run correctly without having to rewrite them all.  People do more than just migrate schema in migrations, and as far as I'm aware it's a properly-supported use case for Rails Migrations (and precisely why everyone finds them so wonderful).

The main fact is that this worked before, and now it doesn't, and your suggestion to (I think) go to SQL statements instead of using the very abstractions that migrations add seems a little counter-productive.

**Comment by akaspick** [comment]
I can't even run a migration with an enum right now if the enum is already declared in my model...
```ruby
class Account
  enum plan_type: [:starter, :standard, :professional]
end
```
```ruby
def change
  change_table :accounts do |t|
    t.integer :plan_type, default: 0
  end
end
```
`Undeclared attribute type for enum 'plan_type'. Enums must be backed by a database column or declared with an explicit type via 'attribute'.`

**Comment by gsar** [comment]
> * There is absolutely no reason to run 3 years old migrations. Use `db:schema:load` to directly start from a recent state.
> 
> Old migrations are meant to be deleted.

@byroot this problem affects 3 week old migrations the same as it does 3 year old migrations, so this is not really an answer to anything. i'm finding myself peppering my concerns with defensive cruft like this now.

``` ruby
model FooBar
    extend ActiveSupport::Concern
    included do
      return unless respond_to?(:foo_bar)
      enum foo_bar: {
        ...
      }
    end
end
```

**Comment by zacheryph** [comment]
@jasonl @arcreative 

Referencing application models in migrations is an anti-pattern. There are reasons for this. For one, you cannot rely on their definition. You have any number of things that will break older migrations; Newer associations will fail old migrations for non-presence for instance. I've broke these kinds of migrations a week after they're created simply by making a column not-null and adding a `validates presence: true`.

As well, you simply cannot rely on their existence. It is not uncommon to rename or even remove a model. Its not _completely_ unfathomable for instance to rename your `User` model to say `Person`? `Post` to `Article`?

When I need (want) a model in a migration, I define one _in_ the migration class that is specific for that migration, with _just_ enough implementation for that migration to succeed. It may not _look_ the best at first but I've simply just made this common practice, and you worry about old migrations failing a _lot_ less. I personally find this _cleaner_ as your migration is then completely self contained.

**Comment by arcreative** [comment]
@zacheryph This is some serious hand-wavery, going to have to agree to disagree here.

1. When adding a new, required association to a model, that association is going to fail whether you use the model class in the migration or not. I'm not sure what this is supposed to highlight, exactly
2. I would expect when you change a model name that you refactor your migrations, just like everything else. It's not any different than replacing any other usage in your codebase. 
3. While creating a temporary model in your migration might have solved the problem for you, having done so with an enum that you considered "_just_ enough implementation" would have broken just the same as it did for us, because they moved the goalposts with the enum class method. I still don't see how your approach fixes anything.

Perhaps I'm not understanding the _why_ of this change--I may have missed it here, but is there rationale/a PR for what was accomplished with this breaking change?

**Comment by gsar** [comment]
@zacheryph let me say this simply. this problem has nothing to do with model dependencies in the migration.
1. add a table with `create_table` in a migration. dont run migration yet.
2. in a second migration, add a column backing a rails `enum`, along with a concern that defines said `enum`.
3. `rails db:migrate` will now throw an exception when running first migration.

**Comment by marzdrel** [comment]
> @zacheryph let me say this simply. this problem has nothing to do with model dependencies in the migration.

To re-iterate on what @zacheryph posted - it's not longer possible to add column and matching enum definition in single PR/change. Application won't boot (model file won't load) to perform the migration if the enum is defined and the column is not there. After this change, adding new column with enum definition requires two separate deployments. 

While separating schema changes from code changes might be a good practice in many cases, there are as many cases when this complicates the flow and prevents fast iterations. This behavior is quite hard to detect early as well. Most developers will iterate in steps: adding columns, then adding enum and most of default pipelines won't be able to detect this change (as the schema is loaded from the schema file on CI servers). 

I don't have a good idea/solution for this issue. Just wanted to make sure that this consequence of the change is clearly stated. 😅

**Comment by synth** [comment]
Agree that this is weird and regressive from my perspective. In our case, we have deploy scripts which are rake tasks that load the Rails env and are set to run before the database migration. These now will crash unless we split the DB changes out which is not a good developer experience, imo. Hoping for a workaround 🤞

**Comment by johan-smits** [comment]
In my case I do use the `anti-pattern` but differently:

```ruby
class AcceptAllAccountsEula < ActiveRecord::Migration[4.2]
  def change
    Account.update_all(eula: false)
  end
end
```

Since this only generates a single query directly into the database, this also fails for the same reason. But it never touches our uses the enum field.

The reason why we keep old migrations in the code and not drop them is that we have customer sites that need to upgrade and thet are not always on the same speed as we are.

Since this is `anti-pattern` what is then the suggested way to do such migrations with Rails?

**Comment by arcreative** [comment]
@johan-smits change `def change` to `def up` and use `execute “UPDATE accounts SET eula = FALSE`.

That said, I think we are all still unhappy about this surprising breaking change despite there being a solution for your particular use case. We shouldn’t have to rewrite all of our migrations (which span many versions!) for what feels like no good reason.

**Comment by johan-smits** [comment]
@arcreative thanks for the example. I was more looking for a "proper" way to do such migrations. But I guess there is nothing in place?

Off to a lot of migration rewrites then...

**Comment by arcreative** [comment]
@johan-smits If I'm understanding correctly, this is what they're suggesting. I have a problem with it in that it forces us to leave the nice abstractions behind that Rails migrations/ActiveModel are meant to provide as a convenience, but they seem to think that I'm wrong about that.

**Comment by gsar** [comment]
@johan-smits what i have resorted to is to precede every enum declaration with an `attribute` line that declares the enum's type. seems to be enough so far, and i havent had to change any migrations. see my [note above](https://github.com/rails/rails/issues/49717#issuecomment-2064889622)

**Comment by johan-smits** [comment]
Currently refactoring some migrations and we use [ClosureTree](https://github.com/ClosureTree/closure_tree) and in one of our migrations we migrated to this gem. But with existing parent data you need to rebuild your tree with `.rebuild!`. This isn't something you can fix with a simple sql query.

I started looking for alternative solutions since reading more about the change they do have a point to change it. [data-migrate](https://github.com/ilyakatz/data-migrate) seems a good alternative for the data manipulation. Will give it a try since they claim tail `7.1` support.

**Comment by ziaulrehman40** [comment]
We are encountering same issue when upgrading an existing application, we have exact named column as well as exact named enum, but when i run `rails db:create db:migrate` it fails, because we try to use a such model in initializers for something, and when initializing, it seemingly breaks. So it i really really annoying and definately worth fixing.

**Comment by ziaulrehman40** [comment]
As for this suggestion https://github.com/rails/rails/issues/49717#issuecomment-1791394231, to use db:schema:load, this doesn't work quiet well when using ros-apartment like gems, so we have to use db:migrate to setup, or when we have things which rails schema doesn't quiet support, and we are using those features of DB, it will not work.

**Comment by vpereira** [comment]
> Since this is `anti-pattern` what is then the suggested way to do such migrations with Rails?

We encountered a similar issue. Our CI runs `rails db:migrate` specifically to ensure we don't leave any broken migration code lying around. One potential solution is to decouple data migration from schema migration using the [data-migrate](https://github.com/ilyakatz/data-migrate) gem. It would be ideal to have an official "Rails way" to address this problem.

**Comment by ziaulrehman40** [comment]
@vpereira Thanks for the suggestion, I ended up cleaning up all my old migrations, and used [squasher](https://github.com/jalkoby/squasher) `squasher 2024 --migration=7.0` to squash my old migrations before 2024(i wanted to do it anyway, to reduce db:migrate time).

**Comment by louiswdavis** [comment]
Like others, we have encountered this issue with historic migrations. We also use the [data-migrate](https://github.com/ilyakatz/data-migrate) gem to separate our schema and data migrations which helps for the most part, but the issue still arises if you call those data migrations.

We'd prefer not to remove/comment out previous data migrations if we can help it so instead decided to capture the error so that the data migrations continue as expected.

```ruby
def up
  data_migration_code
rescue RuntimeError => e
  puts "Error Captured: #{e.full_message}"
end
```

**Comment by marzdrel** [comment]
Leaving the historic migrations aside - it would be great to have a way to add new db columns with accompanying enum definition in a single change. I could easily imagine pushing this even further and require models to have underlying table, which would prevent adding new model in a single change... 

Some sort of opt-out from this requirement to at least run the `db:migrate` task would be a good compromise. 🤔

**Comment by ziaulrehman40** [comment]
BTW we were hitting argumentError, so what we did was add `validate: true` or `validate: { allow_nil: true }` to enums, which adds validations but doesn't raise on wrong or empty assignment. We did still removed old data migrations anyway, as they were not needed.

**Comment by rience** [comment]
I hoped that `reset_column_information` would help here but no luck :(

**Comment by rafaelfranca** [comment]
Reopening since https://github.com/rails/rails/pull/49769 was reverted.

**Comment by duduribeiro** [comment]
Just an information for this. For the implementation for this one we must also test if everything works with parallelized tests when `eager_load` is true. 

See this for more information https://github.com/rails/rails/issues/52607

TLDR; the previous implementation needed to be reverted because it was breaking running tests in parallel while eager_load was set to true. 


We can use this test script provided by @robbevp to test:

```ruby
# frozen_string_literal: true

ENV['RAILS_ENV'] ||= 'test'

require "bundler/inline"

gemfile(true) do
  source "https://rubygems.org"

  gem "rails", "7.2.0"

  gem "pg", "~> 1.5.7"

  gem "minitest", "5.24.1" # https://github.com/minitest/minitest/issues/1007
end

require "active_record"
require "action_controller/railtie"
require "logger"

ActiveRecord::Base.configurations = {
  "test" => {
    "adapter"=>"postgresql",
    "database"=>"test",
    "host"=>"localhost",
    "username"=>"rails",
    "password"=>"rails"
  }
}

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(
  adapter:  "postgresql",
  host:     "localhost",
  username: "rails",
  password: "rails",
  database: "test")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_enum "my_enum", ["one", "two", "three"]

  create_table :posts, force: true do |t|
    t.enum :post_type, enum_type: :my_enum
  end
end


class Application < Rails::Application
  config.eager_load = true
end

Rails.application.initialize!

class Post < ActiveRecord::Base
  enum :post_type, { one: 'one', two: 'two', three: 'three' }
end


class ActiveSupport::TestCase
  parallelize(
    workers: 2,
    threshold: 1
  )
end

require "rails/test_help"

class BugTest < ActiveSupport::TestCase
  def test_one
    pp Post.columns
    Post.new

    assert true
  end

  def test_two
    Post.new

    assert true
  end
end
```

**Comment by shayonj** [comment]
Any chance this can be back ported to rails 7.2 ?

**Comment by duduribeiro** [comment]
> Any chance this can be back ported to rails 7.2 ?

@shayonj you can just use foreman with the proper version when released (or build yourself it) and just updated your Procfile.dev. This is not rails 8 dependent

**Comment by DaAwesomeP** [comment]
Sorry if this has been said, just a bit confused between the 4 issues/pulls. Is it possible to `attr_accessor` an `enum` that is not backed in the DB in Rails 7.2 at the moment?

**Comment by heaven** [comment]
@byroot My issue is with the DB migrations loading classes with enums before a corresponding column was added. Happens in tests, when re-creating the database. Is this also the intended behavior?

**Comment by zzak** [comment]
@heaven Why are you loading classes inside the migration?

**Comment by heaven** [comment]
@zzak I would like to highlight that the problem isn't me doing something that doesn't match someone's preferences or expectations. The actual issue is that the framework fails to load its own components from within the database migration. Most likely because someone has simply forgotten about such a possibility.

This change breaks all migrations that invoke anything on models that have enums, columns for which were added in further migrations which were not invoked just yet.

**Comment by zzak** [comment]
@heaven Sorry, I think my question may have been accusational. What I meant to ask was if you could provide more detail to what it is you're doing, and the expected behavior.

I understand "loading classes with enums before the column was added", but could you give an example?

There was a PR to address this backported to 7.2.2, so if you're experiencing unexpected behavior as a result of this PR -- it may make sense to open a new issue than continue growing this thread.

**Comment by heaven** [comment]
@zzak hi, no problem, sorry for overreacting. I thought that I'd tell I'm updating data from there, then you'd tell me that's not the right place to do so, and we'd derail the conversation nowhere.

I know it is not the best place to do data migration, but I don't agree it is the wrong place to do so. Sometimes you need to update the data within the migration transaction, so if anything goes wrong, then everything is rolled back.

But you don't necessarily need to update the data, it is enough to just preload the model class in there.

Here is an example that should showcase the issue better

```ruby
class User < ApplicationRecord
  enum :status, %w[pending approved]
end

class AddOnboardedToUser < ActiveRecord::Migration
  def change
    add_column :users, :onboarded, :boolean, null: false, default: fasle

    up_only do
      User.unscoped.update_all(onboarded: true)
    end
  end
end

class AddStatusToUser < ActiveRecord::Migration
  def change
    add_column :users, :status, :integer
  end
end
```

You can see two database migrations here, and if the first one was created and executed a long time ago, before the `status` enum was added into the `User` model, everything will be just fine. But if you then decide to do `rake db:migrate:reset` (e.g. to rebuild the test database), the first migration will fail, because the enum is already in the model but there is no corresponding column just yet, it was about to be added in the next migration.

**Comment by akaspick** [comment]
If you're referencing a class in the migration, mock it out in the same migration... see my SO answer at https://stackoverflow.com/questions/77260392/how-can-i-fix-unknown-enum-attribute-after-rails-7-1-upgrade/78041224#78041224

Basically you should be doing something like this...

```ruby
class AddOnboardedToUser < ActiveRecord::Migration
  class User < ApplicationRecord
    attribute :status, :string
  end

  def change
    add_column :users, :onboarded, :boolean, null: false, default: fasle

    up_only do
      User.unscoped.update_all(onboarded: true)
    end
  end
end
```

**Comment by heaven** [comment]
I did so a couple of times in my life, but rather as a hack. I believe this is not the right way to go, not to become a standard approach at least. The case with `#update_all` was just an example, sometimes we need to pull more business logic. In my current project, started almost 10 years ago, around 50 database migrations are failing because of this.

I can also easily see this happening in other situations, e.g. when merging multiple pull-requests, when one adds a migration with data changes and another adds an enum. Where migrations worked alone and will stop working once merged.

I'm pretty much sure this error should have never existed. Honestly, this is none of the framework's business what I do, when, and why. The error could happen when attempting to save the data or using the unum, but not in the runtime during boot. If there is any logic behind that error, that makes it impossible to define an enum without a corresponding column, then this is a regression in comparison with Rails 7 and below, where this was perfectly possible.

In other words, it is just bad design.

**Comment by arianf** [comment]
I feel like when running `db:migrate` missing column validation should just be turned off.
